### ISSUE TITLE
scipy 1.7.0 breaking test_lat2SW

### ISSUE THREAD

**Issue by jGaboardi** [pysal/libpysal#419]
`test_lat2SW` breaking after `scipy 1.7.0` used in testing. Seems to be a change in [`scipy.sparse.dia`](https://github.com/scipy/scipy/commit/23aa0cbaff1754f02acf074006399ba2edede2ef#diff-8a57744fb65927efbe268b2d787bdb06d5b6d4c59a9aa67a924c002604a70460) 

See [here](https://github.com/pysal/libpysal/pull/418#issuecomment-869232551) for more information.

Skipping `test_lat2SW` with `scipy 1.7.0` for now.

See all so #420.

**Comment by jGaboardi** [comment]
[`_add_sparse`](https://github.com/scipy/scipy/blob/97089947eef7351e82718b105c244329d76138ac/scipy/sparse/dia.py#L240) was added in scipy/scipy#14004, which was included in the SciPY 1.7.0 release, seemingly accounting for the change in behavior.

**`scipy 1.6.3`**:
```python
In [1]: import libpysal

In [2]: import scipy

In [3]: scipy.__version__
Out[3]: '1.6.3'

In [4]: libpysal.weights.util.lat2SW(3, 3).nnz
Out[4]: 24
```


**`scipy 1.7.0`**:
```python
In [1]: import libpysal

In [2]: import scipy

In [3]: scipy.__version__
Out[3]: '1.7.0'

In [4]: libpysal.weights.util.lat2SW(3, 3).nnz
Out[4]: 28
```

**Comment by jGaboardi** [comment]
[This gist](https://gist.github.com/jGaboardi/7e115760e3d3d5c02141d6ce90efa6f3) demonstrates the change in behavior occurs within `scipy.sparse.dia_matrix`, not `libpysal.weights.util.lat2SW`, specifically highlighting the difference between `scipy 1.6.3` and `1.7.0` where `_add_sparse` was added.

**Comment by ljwolf** [comment]
Great sleuthing @jGaboardi!

I think the issue is that [scipy.sparse tracks zero as a value](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.nnz.html#scipy.sparse.csc_matrix.nnz) in order to handle cases where we need to record a *structural zero* for the graph algorithms in `scipy.sparse.csgraph`. This means that, when you send a "data" matrix with zeros in it (like we do in `lat2SW`) the sparse matrix tracks that as a "structural" zero. This means it's not counted as a "zero" in the sparse matrix, because that refers to an "empty," basically. `nnz` should really be thought of as `n_nonempty`, not `n_nonzero`. 

The correct fix is to cast the `dia_matrix` to a `csc` and then `csc.eliminate_zeros()`, I think.

**Comment by jGaboardi** [comment]
Being addressed in #426.
